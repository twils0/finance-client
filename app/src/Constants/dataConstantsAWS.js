// eslint-disable-next-line
export const actionTypes = {
  SET_AWS_STATUS: 'DATA_AUTH_AWS_STATUS',
  SET_AWS_USER: 'DATA_AUTH_SET_AWS_USER',
  RESET_AWS_STATE: 'DATA_AUTH_RESET_AWS_STATE',
};
