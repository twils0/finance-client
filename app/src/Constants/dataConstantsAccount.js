export const actionTypes = {
  SET_ACCOUNT_STATUS: 'DATA_ACCOUNT_SET_ACCOUNT_STATUS',
  SET_FIELD: 'DATA_ACCOUNT_SET_FIELD',
  SET_FIELDS: 'DATA_ACCOUNT_SET_FIELDS',
  RESET_ACCOUNT_STATE: 'DATA_ACCOUNT_RESET_ACCOUNT_STATE',
};

export const statusNames = {
  AWS_FIELDS: 'accountStatusAWSFields',
  STRIPE_FIELDS: 'accountStatusStripeFields',
  UPDATE_STRIPE_FIELDS_TOKEN: 'accountStatusUpdateStripeFieldsToken',
  UPDATE_STRIPE_FIELDS_REQUEST: 'accountStatusUpdateStripeFieldsRequest',
  UPDATE_DB_FIELDS: 'accountStatusUpdateDBFields',
  UPDATE_AWS_FIELDS: 'accountStatusUpdateAWSFields',
};

export const fieldNames = {
  NAME: 'accountName',
  EMAIL: 'accountEmail',
  EMAIL_ADDITIONAL: 'accountEmailAdditional',
  PHONE: 'accountPhone',
  NAME_ON_CARD: 'accountNameOnCard',
  PROMO_CODE: 'accountPromoCode',
  PROMO_CODE_VALID: 'accountPromoCodeValid',
  CUSTOMER: 'accountCustomer',
};
