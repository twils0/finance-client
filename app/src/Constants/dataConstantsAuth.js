export const actionTypes = {
  SET_AUTHENTICATED: 'DATA_AUTH_SET_AUTHENTICATED',
  SET_REDIRECT_URL: 'DATA_AUTH_SET_REDIRECT_URL',
  SET_AUTH_STATUS: 'DATA_AUTH_SET_AUTH_STATUS',
  SET_CODE_TYPE: 'DATA_AUTH_SET_CODE_TYPE',
  RESET_AUTH_STATE: 'DATA_AUTH_RESET_AUTH_STATE',
};

export const statusNames = {
  LOGIN: 'authStatusLogin',
  SIGN_OUT_DEVICES: 'authStatusSignOutDevices',
  LOGIN_MFA: 'authStatusLoginMFA',
  SIGN_UP: 'authStatusSignUp',
  VERIFY_PHONE: 'authStatusVerifyPhone',
  VERIFY_PHONE_CODE: 'authStatusVerifyPhoneCode',
  VERIFY_EMAIL: 'authStatusVerifyEmail',
  VERIFY_EMAIL_LINK: 'authStatusVerifyEmailLink',
  FORGOT_PASSWORD: 'authStatusForgotPassword',
  RESET_PASSWORD: 'authStatusResetPassword',
  CHANGE_PASSWORD: 'authStatusChangePassword',
  DELETE_ACCOUNT: 'authStatusDeleteAccount',
  LOGOUT: 'authStatusLogout',
};

export const codeTypeNames = {
  VERIFY_PHONE: 'verifyPhone',
  VERIFY_EMAIL: 'verifyEmail',
  VERIFY_EMAIL_ADDITIONAL: 'verifyEmailAdditional',
};
