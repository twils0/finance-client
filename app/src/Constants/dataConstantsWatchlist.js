export const actionTypes = {
  SET_WATCHLIST_STATUS: 'DATA_WATCHLIST_SECURITIES_STATUS',
  SET_SECURITIES_CURRENT: 'DATA_WATCHLIST_SET_SECURITIES_CURRENT',
  SET_SECURITIES_LIST: 'DATA_WATCHLIST_SET_SECURITIES_LIST',
  SET_SECURITIES_ALL: 'DATA_WATCHLIST_SET_SECURITIES_ALL',
  SET_SECURITY: 'DATA_WATCHLIST_SET_SECURITY',
  SET_SECURITY_DATA: 'DATA_WATCHLIST_SET_SECURITY_DATA',
  SET_SECURITY_SUBSCRIBE: 'DATA_WATCHLIST_SET_SECURITY_SUBSCRIBE',
  RESET_WATCHLIST_STATE: 'DATA_ACCOUNT_RESET_WATCHLIST_STATE',
};

export const codeTypeNames = {
  CONFIRM_PHONE: 'confirmPhone',
  CONFIRM_EMAIL: 'confirmEmail',
};

export const statusNames = {
  SECURITIES: 'watchlistStatusSecurities',
  GET_SECURITY_DATA: 'watchlistStatusGetSecurityData',
  UPDATE_SECURITIES: 'watchlistStatusUpdateSecurities',
};
