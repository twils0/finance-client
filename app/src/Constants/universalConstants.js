import stripeConfig from '../../../stripe.config.json';
import dbConfig from '../../../db.config.json';
import iexConfig from '../../../iex.config.json';

export const baseURLs = {
  STRIPE: stripeConfig.url,
  DB: dbConfig.url,
  IEX: iexConfig.urls.api,
  IEX_DEV: iexConfig.urls.dev,
  IEX_TERMS: iexConfig.urls.terms,
};

export const URLs = {
  CUSTOMERS: '/customers',
  USERS: '/users',
  EMAILS: '/emails',
  SECURITIES: '/securities',
  SECURITIES_SEARCH: '/securities/search',
  IEX_COMPANY: tickerCusip => `/stock/${tickerCusip}/company`,
  IEX_QUOTE: tickerCusip => `/stock/${tickerCusip}/quote`,
};

export const axiosConfig = {
  STRIPE: {
    baseURL: baseURLs.STRIPE,
    timeout: 30000,
  },
  DB: {
    baseURL: baseURLs.DB,
    timeout: 30000,
  },
  IEX: {
    baseURL: baseURLs.IEX,
    timeout: 30000,
  },
};

export const pathNames = {
  DEFAULT: '*',
  LOGIN: '/login',
  SIGN_UP: '/signup',
  CODE: '/code',
  CODE_VERIFY_EMAIL: '/code/:verificationId?',
  TERMS: '/terms',
  INTERNAL: '/internal',
  WATCHLIST: '/internal/watchlist',
  WATCHLIST_SECURITY_ID: '/internal/watchlist/:securityId?',
  ACCOUNT: '/internal/account',
};

export const requestStatusTypes = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR',
};
