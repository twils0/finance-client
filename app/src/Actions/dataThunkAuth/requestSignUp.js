import axios from 'axios';
import raven from 'raven-js';

import handleErrorCatch from '../../handleErrorCatch';
import { requestStatusTypes, URLs, axiosConfig } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsAuth';
import { setAuthStatus } from '../dataActionsAuth';

import { demo } from '../../../../mode.config.json';

const requestSignUp = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'element')) {
    throw new Error(`Please enter a value for the 'element' key - ${JSON.stringify(payload)}`);
  }
  if (typeof payload.element !== 'object' || payload.element.constructor === Array) {
    throw new Error(
      `Please provide an object value for the 'element' key - ${JSON.stringify(payload)}`,
    );
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'nameOnCard')) {
    throw new Error(`Please enter a value for the 'name' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'promoCode')) {
    throw new Error(`Please enter a value for the 'promoCode' key - ${JSON.stringify(payload)}`);
  }

  if (!Object.prototype.hasOwnProperty.call(payload, 'email')) {
    throw new Error(`Please enter a value for the 'email' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'password')) {
    throw new Error(`Please enter a value for the 'password' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'phone')) {
    throw new Error(`Please enter a value for the 'phone' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'name')) {
    throw new Error(`Please enter a value for the 'name' key - ${JSON.stringify(payload)}`);
  }

  const plan = 'basic_plan';

  return async (dispatch, getState) => {
    const state = getState();

    if (state.data.auth.status[statusNames.SIGN_UP].status !== requestStatusTypes.LOADING) {
      dispatch(setAuthStatus({ id: statusNames.SIGN_UP, status: requestStatusTypes.LOADING }));

      try {
        if (!demo) {
          const stripe = state.ui.app.stripe.stripeObject;
          const { token, error } = await stripe.createToken(payload.element, {
            name: payload.name,
          });

          if (error) {
            throw error;
          }

          const signUpPayload = { ...payload };
          signUpPayload.token = token.id;
          signUpPayload.plan = plan;
          delete signUpPayload.element;
          delete signUpPayload.nameOnCard;

          await axios.post(URLs.CUSTOMERS, signUpPayload, axiosConfig.STRIPE);
        }
      } catch (errorCatch) {
        const error = handleErrorCatch(errorCatch);

        dispatch(
          setAuthStatus({
            id: statusNames.SIGN_UP,
            status: requestStatusTypes.ERROR,
          }),
        );

        if (
          typeof error === 'object'
          && error
          && (error.raw
            || error.code === 'UsernameExistsException'
            || error.code === 'coupon_invalid'
            || error.code === 'coupon_expired'
            || error.code === 'card_declined')
        ) {
          return Promise.reject(error);
        }

        if (!demo) {
          raven.captureException(error, {
            logger: 'requestSignUp',
          });
        }

        return Promise.reject(error);
      }

      dispatch(
        setAuthStatus({
          id: statusNames.SIGN_UP,
          status: requestStatusTypes.SUCCESS,
        }),
      );
    }

    return null;
  };
};

export default requestSignUp;
