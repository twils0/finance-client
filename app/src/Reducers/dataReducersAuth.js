import { combineReducers } from 'redux';

import { requestStatusTypes, pathNames } from '../Constants/universalConstants';
import { actionTypes, statusNames, codeTypeNames } from '../Constants/dataConstantsAuth';

export const initialState = {
  authenticated: false,
  redirectURL: pathNames.WATCHLIST,
  status: {
    [statusNames.LOGIN]: {
      id: statusNames.LOGIN,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.SIGN_OUT_DEVICES]: {
      id: statusNames.SIGN_OUT_DEVICES,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.LOGIN_MFA]: {
      id: statusNames.LOGIN_MFA,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.SIGN_UP]: {
      id: statusNames.SIGN_UP,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_PHONE]: {
      id: statusNames.VERIFY_PHONE,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_PHONE_CODE]: {
      id: statusNames.VERIFY_PHONE_CODE,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_EMAIL]: {
      id: statusNames.VERIFY_EMAIL,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.VERIFY_EMAIL_LINK]: {
      id: statusNames.VERIFY_EMAIL_LINK,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.FORGOT_PASSWORD]: {
      id: statusNames.FORGOT_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.RESET_PASSWORD]: {
      id: statusNames.RESET_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.CHANGE_PASSWORD]: {
      id: statusNames.CHANGE_PASSWORD,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.LOGOUT]: {
      id: statusNames.LOGOUT,
      status: requestStatusTypes.SUCCESS,
    },
    [statusNames.DELETE_ACCOUNT]: {
      id: statusNames.DELETE_ACCOUNT,
      status: requestStatusTypes.SUCCESS,
    },
  },
  codeTypes: {
    [codeTypeNames.VERIFY_PHONE]: {
      id: codeTypeNames.VERIFY_PHONE,
      needed: true,
    },
    [codeTypeNames.VERIFY_EMAIL]: {
      id: codeTypeNames.VERIFY_EMAIL,
      needed: true,
    },
    [codeTypeNames.VERIFY_EMAIL_ADDITIONAL]: {
      id: codeTypeNames.VERIFY_EMAIL_ADDITIONAL,
      needed: true,
    },
  },
};

export const authenticated = (state = initialState.authenticated, action) => {
  switch (action.type) {
    case actionTypes.SET_AUTHENTICATED:
      return action.payload.authenticated;
    case actionTypes.RESET_AUTH_STATE:
      return initialState.authenticated;
    default:
      return state;
  }
};

export const redirectURL = (state = initialState.redirectURL, action) => {
  switch (action.type) {
    case actionTypes.SET_REDIRECT_URL:
      return action.payload.redirectURL;
    default:
      return state;
  }
};

export const status = (state = initialState.status, action) => {
  switch (action.type) {
    case actionTypes.SET_AUTH_STATUS:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    case actionTypes.RESET_AUTH_STATE:
      return initialState.status;
    default:
      return state;
  }
};

export const codeTypes = (state = initialState.codeTypes, action) => {
  switch (action.type) {
    case actionTypes.SET_CODE_TYPE:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    case actionTypes.RESET_AUTH_STATE:
      return initialState.codeTypes;
    default:
      return state;
  }
};

const dataReducersAuth = combineReducers({
  authenticated,
  redirectURL,
  status,
  codeTypes,
});

export default dataReducersAuth;
