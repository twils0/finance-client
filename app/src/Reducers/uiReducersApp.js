import { combineReducers } from 'redux';

import stripeConfig from '../../../stripe.config.json';
import { requestStatusTypes } from '../Constants/universalConstants';
import { actionTypes, imageNames } from '../Constants/uiConstantsApp';
import {
  formNames as formNamesSignUp,
  inputNames as inputNamesSignUp,
} from '../Constants/uiConstantsExternal';
import {
  formNames as formNamesAccount,
  inputNames as inputNamesAccount,
} from '../Constants/uiConstantsAccount';

import exampleHeader from '../Images/exampleHeader.png';

export const initialState = {
  fonts: {
    status: requestStatusTypes.IDLE,
  },
  images: {
    list: [...Object.values(imageNames)],
    [imageNames.EXAMPLE_HEADER]: {
      id: imageNames.EXAMPLE_HEADER,
      src: exampleHeader,
      status: requestStatusTypes.IDLE,
    },
  },
  stripe: {
    apiKey: stripeConfig.apiKey,
    status: requestStatusTypes.IDLE,
    elements: {
      [inputNamesSignUp[formNamesSignUp.CARD].STRIPE]: {
        id: inputNamesSignUp[formNamesSignUp.CARD].STRIPE,
        loaded: false,
      },
      [inputNamesAccount[formNamesAccount.BILLING].STRIPE]: {
        id: inputNamesAccount[formNamesAccount.BILLING].STRIPE,
        loaded: false,
      },
    },
    stripeObject: null,
  },
};

export const fonts = (state = initialState.fonts, action) => {
  switch (action.type) {
    case actionTypes.SET_FONTS_STATUS:
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
};

export const images = (state = initialState.images, action) => {
  switch (action.type) {
    case actionTypes.SET_IMAGE_STATUS:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    default:
      return state;
  }
};

export const stripe = (state = initialState.stripe, action) => {
  switch (action.type) {
    case actionTypes.SET_STRIPE_STATUS:
      return {
        ...state,
        ...action.payload,
      };
    case actionTypes.SET_STRIPE_ELEMENT_LOADED:
      return {
        ...state,
        elements: {
          ...state.elements,
          [action.payload.id]: {
            ...state.elements[action.payload.id],
            ...action.payload,
          },
        },
      };
    case actionTypes.SET_STRIPE_OBJECT:
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
};

const uiReducersApp = combineReducers({
  fonts,
  images,
  stripe,
});

export default uiReducersApp;
