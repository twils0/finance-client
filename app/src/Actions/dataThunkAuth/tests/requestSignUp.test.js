import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import axios from 'axios';

import { initialState as initialStateAuth } from '../../../Reducers/dataReducersAuth';
import { initialState as initialStateUIApp } from '../../../Reducers/uiReducersApp';
import { requestStatusTypes, URLs, axiosConfig } from '../../../Constants/universalConstants';
import { actionTypes as actionTypesAuth, statusNames } from '../../../Constants/dataConstantsAuth';
import requestSignUp from '../requestSignUp';

jest.mock('axios', () => ({
  post: jest.fn(),
}));

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

const tokenId = 'testTokenId';
const plan = 'basic_plan';
const postPayload = {
  promoCode: 'testPromo',
  email: 'testEmail',
  password: 'testPassword1!',
  phone: '+12395550000',
  name: 'testName',
  token: tokenId,
  plan,
};
const signUpPayload = {
  ...postPayload,
  element: { test: 'testElement' },
  nameOnCard: 'testNameOnCard',
};

describe('dataThunkAuth', () => {
  describe('requestSignUp', () => {
    afterEach(() => {
      axios.post.mockReset();
    });

    it("fails and throws an error when missing 'element' key in payload", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      const emptyPayload = {};

      try {
        await store.dispatch(requestSignUp(emptyPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(axios.post).not.toBeCalled();
    });

    it("fails when provided a payload with an 'element' key that is not an object", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      axios.post.mockReturnValue(Promise.resolve());

      const wrongPayload = { element: 'testElement' };

      try {
        await store.dispatch(requestSignUp(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(axios.post).not.toBeCalled();
    });

    it("fails and throws an error when missing 'nameOnCard' key in payload", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      const wrongPayload = { element: { test: 'testElement' } };

      try {
        await store.dispatch(requestSignUp(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(axios.post).not.toBeCalled();
    });

    it("fails and throws an error when missing 'promoCode' key in payload", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      const wrongPayload = { element: { test: 'testElement' }, nameOnCard: 'testNameOnCard' };

      try {
        await store.dispatch(requestSignUp(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(axios.post).not.toBeCalled();
    });

    it("fails and throws an error when missing 'email' key in payload", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      const wrongPayload = {
        element: { test: 'testElement' },
        nameOnCard: 'testNameOnCard',
        promoCode: 'testPromoCode',
      };

      try {
        await store.dispatch(requestSignUp(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(axios.post).not.toBeCalled();
    });

    it("fails and throws an error when missing 'password' key in payload", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      axios.post.mockReturnValue(Promise.resolve());

      const wrongPayload = {
        element: { test: 'testElement' },
        nameOnCard: 'testNameOnCard',
        promoCode: 'testPromoCode',
        email: 'testEmail',
      };

      try {
        await store.dispatch(requestSignUp(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(axios.post).not.toBeCalled();
    });

    it("fails and throws an error when missing 'phone' key in payload", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      axios.post.mockReturnValue(Promise.resolve());

      const wrongPayload = {
        element: { test: 'testElement' },
        nameOnCard: 'testNameOnCard',
        promoCode: 'testPromoCode',
        email: 'testEmail',
        password: 'testPassword1!',
      };

      try {
        await store.dispatch(requestSignUp(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(axios.post).not.toBeCalled();
    });

    it("fails and throws an error when missing 'name' key in payload", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      axios.post.mockReturnValue(Promise.resolve());

      const wrongPayload = {
        element: { test: 'testElement' },
        nameOnCard: 'testNameOnCard',
        promoCode: 'testPromoCode',
        email: 'testEmail',
        password: 'testPassword1!',
        phone: '+12395550000',
      };

      try {
        await store.dispatch(requestSignUp(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(axios.post).not.toBeCalled();
    });

    it('fails and returns promise reject when stripe.createToken throws an error', async () => {
      const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
      const tokenResult = {
        token: { id: tokenId },
        error: { code: 'testCode', message: 'testMessage' },
      };
      const createToken = jest.fn();
      createToken.mockReturnValue(Promise.resolve(tokenResult));
      stateBeforeUIApp.stripe.stripeObject = { createToken };

      const store = mockStore({ data: { auth: initialStateAuth }, ui: { app: stateBeforeUIApp } });

      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_UP,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_UP,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      try {
        await store.dispatch(requestSignUp(signUpPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(tokenResult.error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(axios.post).not.toBeCalled();
    });

    it('fails and returns promise reject when axios.post throws an error', async () => {
      const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
      const tokenResult = { token: { id: tokenId } };
      const createToken = jest.fn();
      createToken.mockReturnValue(Promise.resolve(tokenResult));
      stateBeforeUIApp.stripe.stripeObject = { createToken };

      const store = mockStore({ data: { auth: initialStateAuth }, ui: { app: stateBeforeUIApp } });

      const error = { code: 'testError', message: 'testMessage' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_UP,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_UP,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      axios.post.mockReturnValue(Promise.reject(error));

      try {
        await store.dispatch(requestSignUp(signUpPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(axios.post).toBeCalledWith(URLs.CUSTOMERS, postPayload, axiosConfig.STRIPE);
    });

    it('creates the correct actions with the correct payloads', async () => {
      const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
      const tokenResult = { token: { id: tokenId } };
      const createToken = jest.fn();
      createToken.mockReturnValue(Promise.resolve(tokenResult));
      stateBeforeUIApp.stripe.stripeObject = { createToken };

      const store = mockStore({ data: { auth: initialStateAuth }, ui: { app: stateBeforeUIApp } });

      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_UP,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.SIGN_UP,
            status: requestStatusTypes.SUCCESS,
          },
        },
      ];

      axios.post.mockReturnValue(Promise.resolve());

      const result = await store.dispatch(requestSignUp(signUpPayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
      expect(axios.post).toBeCalledWith(URLs.CUSTOMERS, postPayload, axiosConfig.STRIPE);
    });
  });
});
