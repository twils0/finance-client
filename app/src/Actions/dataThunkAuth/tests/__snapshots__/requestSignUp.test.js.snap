// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataThunkAuth requestSignUp fails and throws an error when missing 'element' key in payload 1`] = `[Error: Please enter a value for the 'element' key - {}]`;

exports[`dataThunkAuth requestSignUp fails and throws an error when missing 'email' key in payload 1`] = `[Error: Please enter a value for the 'email' key - {"element":{"test":"testElement"},"nameOnCard":"testNameOnCard","promoCode":"testPromoCode"}]`;

exports[`dataThunkAuth requestSignUp fails and throws an error when missing 'name' key in payload 1`] = `[Error: Please enter a value for the 'name' key - {"element":{"test":"testElement"},"nameOnCard":"testNameOnCard","promoCode":"testPromoCode","email":"testEmail","password":"testPassword1!","phone":"+12395550000"}]`;

exports[`dataThunkAuth requestSignUp fails and throws an error when missing 'nameOnCard' key in payload 1`] = `[Error: Please enter a value for the 'name' key - {"element":{"test":"testElement"}}]`;

exports[`dataThunkAuth requestSignUp fails and throws an error when missing 'password' key in payload 1`] = `[Error: Please enter a value for the 'password' key - {"element":{"test":"testElement"},"nameOnCard":"testNameOnCard","promoCode":"testPromoCode","email":"testEmail"}]`;

exports[`dataThunkAuth requestSignUp fails and throws an error when missing 'phone' key in payload 1`] = `[Error: Please enter a value for the 'phone' key - {"element":{"test":"testElement"},"nameOnCard":"testNameOnCard","promoCode":"testPromoCode","email":"testEmail","password":"testPassword1!"}]`;

exports[`dataThunkAuth requestSignUp fails and throws an error when missing 'promoCode' key in payload 1`] = `[Error: Please enter a value for the 'promoCode' key - {"element":{"test":"testElement"},"nameOnCard":"testNameOnCard"}]`;

exports[`dataThunkAuth requestSignUp fails when provided a payload with an 'element' key that is not an object 1`] = `[Error: Please provide an object value for the 'element' key - {"element":"testElement"}]`;
